node/libs/protobuf_build/src/syntax.rs

```diff
@@ -0,0 +1,244 @@
+        self
+    }
+}
+
+impl fmt::Display for RustName {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str(&self.0.join("::"))
+    }
+}
+
+impl From<prost_build::Module> for RustName {
+    fn from(s: prost_build::Module) -> Self {
+        Self(s.parts().map(Part::from).collect())
+    }
+}
+
+impl From<&str> for RustName {
+    fn from(s: &str) -> Self {
+        Self(s.split("::").map(Part::from).collect())
+    }
+}
+
+/// A rust module representation.
+/// It is used to collect the generated protobuf code.
+#[derive(Default)]
+pub(super) struct RustModule {
+    /// Nested modules which transitively contain the generated code.
+    modules: BTreeMap<Part, RustModule>,
+    /// Code of the module.
+    code: String,
+}
+
+impl RustModule {
+    /// Returns a reference to a given submodule.
+    pub(crate) fn sub(&mut self, path: &RustName) -> &mut Self {
+        let mut m = self;
+        for part in &path.0 {
+            m = m.modules.entry(part.into()).or_default();
+        }
+        m
+    }
+
+    /// Appends code to the module.
+    pub(crate) fn append(&mut self, code: &str) {
+        self.code += code;
+    }
+
+    fn collect(&self) -> String {
+        let mut entries = vec![self.code.clone()];
+        entries.extend(
+            self.modules
+                .iter()
+                .map(|(name, m)| format!("pub mod {name} {{ {} }}\n", m.collect())),
+        );
+        entries.join("")
+    }
+
+    /// Collects the code of the module and formats it.
+    pub(crate) fn format(&self) -> anyhow::Result<String> {
+        Ok(prettyplease::unparse(
+            &syn::parse_str(&self.collect()).context("syn::parse_str()")?,
+        ))
+    }
+}
+
+/// In addition to input paths and proto paths, there are also proto names
+/// which are used to reference message types from different proto files.
+/// Theoretically proto names can be totally independent from the proto paths (i.e. you can have
+/// "my.favorite.package" proto package defined in "totally/unrelated/file/name.proto" file, but we
+/// recommend the following naming convention: proto package "a.b.c" should be defined either:
+/// a) in a single file "a/b/c.proto", or
+/// b) in a collection of files under "a/b/c/" directory
+/// Option b) is useful for defining large packages, because there is no equivalent of "pub use" in proto syntax.
+#[derive(Clone, PartialEq, Eq)]
+pub struct ProtoName(Vec<String>);
+
+impl ProtoName {
+    /// Checks if package path starts with the given prefix.
+    pub fn contains(&self, elem: &Self) -> bool {
+        elem.0.len() >= self.0.len() && elem.0[0..self.0.len()] == self.0
+    }
+
+    /// Strips a given prefix from the name.
+    pub fn relative_to(&self, prefix: &Self) -> anyhow::Result<Self> {
+        if !prefix.contains(self) {
+            anyhow::bail!(
+                "{} does not contain {}",
+                self.to_string(),
+                prefix.to_string()
```

Comment:
Nit: `to_string()` calls are redundant for types implementing `Display`.

Reply:
done